File: argus/embed.py
"""Ollama embedding client, shared by knowledge packs and the private index.

The model and dimension here are the ones a pack records in its metadata and
that ``packs.format.require_compatible`` refuses to serve a mismatch against.
Changing either constant invalidates every pack already built.

This module's real job is refusing to half-succeed. A pack is built once and
distributed; an embedding client that returns a short or misaligned batch
produces a pack that builds cleanly, queries cleanly, and returns the wrong
documents forever. Every response is therefore checked for count, dimension
and degeneracy before any vector is returned, and a failure anywhere discards
the whole call rather than handing back the batches that did succeed.
"""

from __future__ import annotations

import json
import math
import os

import httpx

EMBED_MODEL = "nomic-embed-text"
EMBED_DIM = 768

# Ollama accepts an arbitrarily long input list, but a local model on CPU can
# take minutes over one, and a timeout mid-way costs the whole request. Batches
# keep each request's worst case bounded.
BATCH_SIZE = 64

DEFAULT_BASE_URL = "http://localhost:11434"

# Generous: first call after a cold start pays for loading the model into
# memory, which on CPU can take tens of seconds before any token is produced.
TIMEOUT = 120.0


class EmbeddingUnavailable(RuntimeError):
    """Embeddings could not be produced, so no embeddings are returned.

    Raised for transport failures, error responses, and responses that arrive
    intact but cannot be trusted -- wrong count, wrong dimension, or a zero
    vector. Callers must treat this as "the pack cannot be built" rather than
    retrying past it, because the alternative is a silently incomplete pack.
    """


def embed_batch(
    texts: list[str], *, client: httpx.Client | None = None,
    base_url: str | None = None,
) -> list[list[float]]:
    """Embed ``texts``, returning one L2-normalized vector per input, in order.

    Normalization is applied even though nothing downstream currently requires
    it: ``quantize.to_bits`` reads signs, ``quantize.to_int8`` scales
    per-vector, and ``quantize.rescore`` computes cosine, all of which are
    invariant under a positive scalar. It is done because a pack is a shared
    artifact -- other tooling, and any future path that lets sqlite-vec compute
    an L2 distance itself, will assume unit vectors, and nomic-embed-text does
    not return them normalized. The invariant is pinned by a test in
    ``tests/test_embed.py``, so a quantiser cannot quietly acquire a scale
    dependence without something failing.
    """
    if not texts:
        return []

    base = (base_url or os.environ.get("ARGUS_OLLAMA_URL")
            or DEFAULT_BASE_URL).rstrip("/")
    owns_client = client is None
    client = client or httpx.Client(timeout=TIMEOUT)

    vectors: list[list[float]] = []
    try:
        for start in range(0, len(texts), BATCH_SIZE):
            batch = texts[start:start + BATCH_SIZE]
            vectors.extend(_embed_one(client, base, batch))
    finally:
        if owns_client:
            client.close()
    return vectors


def _embed_one(
    client: httpx.Client, base: str, batch: list[str]
) -> list[list[float]]:
    try:
        response = client.post(
            f"{base}/api/embed",
            json={"model": EMBED_MODEL, "input": batch},
        )
    except httpx.HTTPError as exc:
        raise EmbeddingUnavailable(
            f"POST {base}/api/embed failed: {exc}"
        ) from exc

    if response.status_code != 200:
        raise EmbeddingUnavailable(
            f"POST /api/embed returned {response.status_code}: "
            f"{response.text[:200]}"
        )

    try:
        body = response.json()
    except json.JSONDecodeError as exc:
        raise EmbeddingUnavailable(
            f"POST /api/embed: failed to decode JSON: {response.text[:200]}"
        ) from exc

    raw = body.get("embeddings")
    if raw is None:
        # Ollama reports a missing or unpulled model as a 200 with an error
        # body, so the message has to carry it or the operator sees nothing
        # actionable.
        raise EmbeddingUnavailable(
            f"POST /api/embed returned no embeddings: {body.get('error', body)}"
        )

    if len(raw) != len(batch):
        # The silent-corruption case: a short batch would shift every later
        # chunk onto its neighbour's vector.
        raise EmbeddingUnavailable(
            f"POST /api/embed: expected {len(batch)} embeddings, got "
            f"{len(raw)} -- refusing to return a misaligned batch"
        )

    return [_normalise(vec, index) for index, vec in enumerate(raw)]


def _normalise(vec: list[float], index: int) -> list[float]:
    if len(vec) != EMBED_DIM:
        raise EmbeddingUnavailable(
            f"embedding {index} has {len(vec)} dimensions, expected "
            f"{EMBED_DIM} -- is {EMBED_MODEL!r} the model actually serving?"
        )

    norm = math.sqrt(sum(x * x for x in vec))
    if norm == 0.0:
        raise EmbeddingUnavailable(
            f"embedding {index} is the zero vector, which has no direction "
            f"and would match everything and nothing"
        )
    return [x / norm for x in vec]

File: dotnet/src/Argus/Indexing/Embed.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Argus.Indexing;

/// <summary>The embedding service could not produce vectors.</summary>
public sealed class EmbeddingUnavailable(string message, Exception? inner = null) : Exception(message, inner);

/// <summary>
/// Text to unit vectors through an OpenAI-compatible <c>/v1/embeddings</c>
/// endpoint -- in the stack, a llama.cpp server running nomic-embed-text.
///
/// Every vector is L2-normalised here, so cosine is a dot product downstream and
/// a zero vector -- which has no direction -- is refused rather than stored.
/// </summary>
public static class Embed
{
    public static string Model => Environment.GetEnvironmentVariable("ARGUS_EMBED_MODEL") is { Length: > 0 } m ? m : "nomic-embed-text";
    public static int Dim => int.TryParse(Environment.GetEnvironmentVariable("ARGUS_EMBED_DIM"), out var d) ? d : 768;
    public const int BatchSize = 64;
    public const string DefaultBaseUrl = "http://localhost:8081";
    static readonly TimeSpan Timeout = TimeSpan.FromSeconds(120);

    /// <summary>Where the embedding server lives: <c>ARGUS_EMBED_URL</c>, without the <c>/v1</c>.</summary>
    public static string BaseUrl =>
        (Environment.GetEnvironmentVariable("ARGUS_EMBED_URL") is { Length: > 0 } u ? u : DefaultBaseUrl).TrimEnd('/');

    /// <summary>Swappable for tests; the production embedder calls the embedding server.</summary>
    public static Func<IReadOnlyList<string>, List<double[]>>? Override { get; set; }

    static readonly Lazy<HttpClient> SharedClient = new(() => new HttpClient { Timeout = Timeout });

    public static List<double[]> EmbedBatch(IReadOnlyList<string> texts, HttpClient? client = null, string? baseUrl = null)
    {
        if (texts.Count == 0) return [];
        if (Override is not null) return Override(texts);
        var b = (baseUrl ?? BaseUrl).TrimEnd('/');
        client ??= SharedClient.Value;
        var vectors = new List<double[]>(texts.Count);
        for (int start = 0; start < texts.Count; start += BatchSize)
        {
            var batch = texts.Skip(start).Take(BatchSize).ToList();
            vectors.AddRange(EmbedOne(client, b, batch));
        }
        return vectors;
    }

    static List<double[]> EmbedOne(HttpClient client, string b, List<string> batch)
    {
        HttpResponseMessage response;
        string text;
        var url = $"{b}/v1/embeddings";
        try
        {
            var payload = new JsonObject { ["model"] = Model, ["input"] = new JsonArray(batch.Select(t => (JsonNode?)t).ToArray()) };
            // A sized body, not JsonContent's chunked stream: some proxies reject a
            // request with no Content-Length.
            using var request = new HttpRequestMessage(HttpMethod.Post, url)
            {
                Content = new StringContent(payload.ToJsonString(), System.Text.Encoding.UTF8, "application/json"),
            };
            if (Environment.GetEnvironmentVariable("ARGUS_EMBED_API_KEY") is { Length: > 0 } key)
                request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", key);
            response = client.Send(request);
            text = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
        }
        catch (Exception exc) when (exc is HttpRequestException or TaskCanceledException or IOException)
        {
            throw new EmbeddingUnavailable($"POST {url} failed: {exc.Message}", exc);
        }
        if ((int)response.StatusCode != 200)
            throw new EmbeddingUnavailable($"POST /v1/embeddings returned {(int)response.StatusCode}: {Util.PyStr.Prefix(text, 200)}");
        JsonNode? body;
        try { body = JsonNode.Parse(text); }
        catch (JsonException exc) { throw new EmbeddingUnavailable($"POST /v1/embeddings: failed to decode JSON: {Util.PyStr.Prefix(text, 200)}", exc); }
        if (body is not JsonObject obj || obj["data"] is not JsonArray data)
        {
            var error = (body as JsonObject)?["error"]?.ToJsonString() ?? body?.ToJsonString() ?? "null";
            throw new EmbeddingUnavailable($"POST /v1/embeddings returned no embeddings: {error}");
        }
        if (data.Count != batch.Count)
            throw new EmbeddingUnavailable(
                $"POST /v1/embeddings: expected {batch.Count} embeddings, got {data.Count} -- refusing to return a misaligned batch");
        // The protocol carries an index per item; order by it rather than trusting
        // the array order, so a server that answers out of order cannot misalign.
        var ordered = data.Select((item, i) => (Index: (item as JsonObject)?["index"]?.GetValue<int>() ?? i, Item: item))
            .OrderBy(x => x.Index).ToList();
        var result = new List<double[]>(ordered.Count);
        for (int i = 0; i < ordered.Count; i++)
        {
            var vec = ((ordered[i].Item as JsonObject)?["embedding"] as JsonArray ?? []).Select(x => x!.GetValue<double>()).ToArray();
            result.Add(Normalise(vec, i));
        }
        return result;
    }

    public static double[] Normalise(double[] vec, int index)
    {
        if (vec.Length != Dim)
            throw new EmbeddingUnavailable(
                $"embedding {index} has {vec.Length} dimensions, expected {Dim} -- is '{Model}' the model actually serving?");
        double norm = 0;
        foreach (var x in vec) norm += x * x;
        norm = Math.Sqrt(norm);
        if (norm == 0)
            throw new EmbeddingUnavailable($"embedding {index} is the zero vector, which has no direction and would match everything and nothing");
        return vec.Select(x => x / norm).ToArray();
    }
}
